xarray/core/variable.py

```diff
@@ -211,7 +211,11 @@ def as_compatible_data(data, fastpath=False):
     if isinstance(data, (Variable, DataArray)):
         return data.data
 
-    if isinstance(data, NON_NUMPY_SUPPORTED_ARRAY_TYPES):
+    if (
+        isinstance(data, NON_NUMPY_SUPPORTED_ARRAY_TYPES)
+        or hasattr(data, "__array_function__")
```

Comment:
> There's an `__array_function__` check down below (lines 245, so it seems like this change should go there).

Good point - thanks for the suggestion! I've moved the change to down below.

> I suspect we can also delete cupy, dask from `NON_NUMPY_SUPPORTED_ARRAY_TYPES`

I haven't tried this. Sounds like this would be a separate change?